src/libcore/num/cast.rs

```diff
@@ -0,0 +1,86 @@
+use fmt;
+
+/// Internal trait for APIs like `i32::cast`.
+#[unstable(feature = "num_cast_internals", issue = "0")]
+pub trait Cast<T>: Sized {
+    #[doc(hidden)]
+    fn cast(t: T) -> Result<Self, CastError>;
+}
+
+/// Error type returned from APIs like `i32::cast`, indicates that a cast could
+/// not be performed losslessly.
+#[unstable(feature = "num_cast", issue = "0")]
+#[derive(Debug)]
+pub struct CastError(());
+
+#[unstable(feature = "num_cast", issue = "0")]
+impl fmt::Display for CastError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        "failed to losslessly cast integral types".fmt(f)
+    }
+}
+
+macro_rules! same_sign_cast_int_impl {
+    ($storage:ty, $target:ty, $($source:ty),*) => {$(
+        #[unstable(feature = "num_cast", issue = "0")]
+        impl Cast<$source> for $target {
+            fn cast(u: $source) -> Result<$target, CastError> {
+                let min = <$target>::min_value() as $storage;
+                let max = <$target>::max_value() as $storage;
+                if u as $storage < min || u as $storage > max {
+                    Err(CastError(()))
+                } else {
+                    Ok(u as $target)
+                }
+            }
+        }
+    )*}
+}
+
+same_sign_cast_int_impl!(u128, u8, u8, u16, u32, u64, u128, usize);
+same_sign_cast_int_impl!(i128, i8, i8, i16, i32, i64, i128, isize);
+same_sign_cast_int_impl!(u128, u16, u8, u16, u32, u64, u128, usize);
+same_sign_cast_int_impl!(i128, i16, i8, i16, i32, i64, i128, isize);
+same_sign_cast_int_impl!(u128, u32, u8, u16, u32, u64, u128, usize);
+same_sign_cast_int_impl!(i128, i32, i8, i16, i32, i64, i128, isize);
+same_sign_cast_int_impl!(u128, u64, u8, u16, u32, u64, u128, usize);
+same_sign_cast_int_impl!(i128, i64, i8, i16, i32, i64, i128, isize);
+same_sign_cast_int_impl!(u128, u128, u8, u16, u32, u64, u128, usize);
+same_sign_cast_int_impl!(i128, i128, i8, i16, i32, i64, i128, isize);
+same_sign_cast_int_impl!(u128, usize, u8, u16, u32, u64, u128, usize);
+same_sign_cast_int_impl!(i128, isize, i8, i16, i32, i64, i128, isize);
+
+macro_rules! cross_sign_cast_int_impl {
+    ($unsigned:ty, $($signed:ty),*) => {$(
+        #[unstable(feature = "num_cast", issue = "0")]
+        impl Cast<$unsigned> for $signed {
+            fn cast(u: $unsigned) -> Result<$signed, CastError> {
+                let max = <$signed>::max_value() as u128;
```

Comment:
Instead of using u128 here I think we should rather implement a (non public) trait `Signed` for each unsigned type that has an assoc type with the corresponding unsigned type, and then perform conversion via `<$signed>::Unsigned`. At least as long as we can't prove that the u128 conversion is optimized away.